### PR TITLE
Creating a Hand Gesture Volume Controller using Python and Pycharm

### LINKED ISSUE #5112: (Python) Creating A Hand Gesture Volume Controller Using Python And Pycharm
### Topic Suggestion
(Python) Creating A Hand Gesture Volume Controller Using Python And Pycharm

### Writing sample(s): 
Include any links or writing samples - to help our team better gauge your writing quality.

#### Proposed title of article
Creating a hand gesture volume controller using python and pycharm

#### Proposed article introduction
In this technological era, computers are crucial to our lives. We use them for working on documents, coding, entertainment, etc. These devices are also easily damaged by water and dust. We therefore cannot handle them with dirty or wet hands. When using them for entertainment, we wash or dry our hands before touching the computer e.g while increasing the volume of our favorite song. This interrupts us from whatever we were doing.

Python helps us solve this issue by enabling us to create a hand gesture volume controller. This software will enable us to control the volume level of our computer by using a hand gesture. We, therefore, do not need to be hands-on on our computers.

In this tutorial, we are going to learn how to use gesture control to change the volume of a computer. We are also going to explain in detail the python packages we will require. We will use the following packages:

- numpy
- cv2
- mediapipe
- math
- time
- ctypes
- comtypes
- pycaw

#### Key takeaways 
The reader will be able to:

- Have an in-depth understanding of how the above packages are used.
- Use python to track hand gestures.
- Create a hand gesture volume controller.

#### Article quality
Unlike other articles, the code will be well explained and not just copy-pasted. The article will also cover in detail the python packages used and why we are using them not just installing them in our project.

#### References
N/A

### PR DESCRIPTION
**Pull request by leeseshawn**
### Article checklist

> If you have not already, please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve your overall technical writing so reviews are swifter, increase the chance of a payout and provide more value to our readers.

> For first-time contributors and for more details on our submission guidelines, **see our [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**.

#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is **600x400 pixels**, under **300KB in size**, and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

This closes #5112

### CODE REVIEW

**Comment by leeseshawn** [comment]
@Bayler 
This is a duplicate of #5403

File: content/articles/creating-a-hand-gesture-volume-controller-using-python-and-pycharm/index.md [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+```python
+cv2.line(img, (x1, y1), (x2, y2), (255, 0, 0), 3)
+```
+In the code above, we use the `cv2.line` function to draw a line between point four of the hand and point `8`. The line will connect point `4` (`(x1, y1)`) which is the tip of the thumb, and point `8` (`(x2, y2)`) which is the tip of the index finger. `(255, 0, 0)` is the color of the line and `3` is its thickness.
+
+### Step 11: Finding the distance between points 4 and 8
+```python
+length = hypot(x2 - x1, y2 - y1)
+```
+In the code above, we find the distance between the tip of the thumb and the index finger using hypotenuse. We achieve this by calling the math `hypot` function and passing the difference between `x2` and `x1` and the difference between `y2` and `y1`.
```
Comment (Bayler) [inline content/articles/creating-a-hand-gesture-volume-controller-using-python-and-pycharm/index.md:199]:
```suggestion
In the code above, we find the distance between the tip of the thumb and the index finger using a hypotenuse. We achieve this by calling the math `hypot` function and passing the difference between `x2` and `x1` and the difference between `y2` and `y1`.
```

File: content/articles/creating-a-hand-gesture-volume-controller-using-python-and-pycharm/index.md [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+The above code draws a circle at the tip of the thumb and at the tip of the index finger.
+- ` (x1, y1)`specifies that we will draw the circle at the tip of the thumb. `15`is the *radius* of the circle. `(255, 0, 0)` is the *color* of the circle. `cv2.FILLED` refers to the thickness of `-1` pixels which will fill the circle with the color we specify.
+
+- We will repeat the same for the index finger in the line of code: 
+
+### Step 10: Drawing a line between points 4 and 8
+```python
+cv2.line(img, (x1, y1), (x2, y2), (255, 0, 0), 3)
+```
+In the code above, we use the `cv2.line` function to draw a line between point four of the hand and point `8`. The line will connect point `4` (`(x1, y1)`) which is the tip of the thumb, and point `8` (`(x2, y2)`) which is the tip of the index finger. `(255, 0, 0)` is the color of the line and `3` is its thickness.
```
Comment (Bayler) [inline content/articles/creating-a-hand-gesture-volume-controller-using-python-and-pycharm/index.md:193]:
```suggestion
In the code above, we use the `cv2.line` function to draw a line between point four of the hand and point `8`. The line will connect point `4` (`(x1, y1)`), which is the tip of the thumb, and point `8` (`(x2, y2)`), which is the tip of the index finger. `(255, 0, 0)` is the line color and `3` is its thickness.
```

File: content/articles/creating-a-hand-gesture-volume-controller-using-python-and-pycharm/index.md [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+
+Refer to the hand image diagram we discussed to identify the [points](#referral-hand-image).
+
+### Step 9: Drawing a circle between the tip of the thumb and the tip of the index finger
+```python                                            
+       cv2.circle(img, (x1, y1), 15, (255, 0, 0), cv2.FILLED)  
+       cv2.circle(img, (x2, y2), 15, (255, 0, 0), cv2.FILLED)  
+```
+The above code draws a circle at the tip of the thumb and at the tip of the index finger.
+- ` (x1, y1)`specifies that we will draw the circle at the tip of the thumb. `15`is the *radius* of the circle. `(255, 0, 0)` is the *color* of the circle. `cv2.FILLED` refers to the thickness of `-1` pixels which will fill the circle with the color we specify.
```
Comment (Bayler) [inline content/articles/creating-a-hand-gesture-volume-controller-using-python-and-pycharm/index.md:185]:
```suggestion
The above code draws a circle at the tip of the thumb and the tip of the index finger.
- ` (x1, y1)` specifies that we will draw the circle at the tip of the thumb. `15`is the *radius* of the circle. `(255, 0, 0)` is the *color* of the circle. `cv2.FILLED` refers to the thickness of `-1` pixels which will fill the circle with the color we specify.
```

File: content/articles/creating-a-hand-gesture-volume-controller-using-python-and-pycharm/index.md [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+### Step 7: Creating a for loop to manipulate each hand
+```python
+for handlandmark in results.multi_hand_landmarks:
+    for id, lm in enumerate(handlandmark.landmark):
+        h, w, c = img.shape
+        cx, cy = int(lm.x * w), int(lm.y * h)
+        lmList.append([id, cx, cy]) 
+    mpDraw.draw_landmarks(img, handlandmark, mpHands.HAND_CONNECTIONS)
+```
+- In the code above, we are using the first for loop to interact with each hand in the results. We are then using the second for loop to get the `id` (id number) and `lm` (landmark information) for each handlandmark. The landmark information will give us the `x` and `y` coordinates. The id number is the number assigned to the various hand points.
```
Comment (Bayler) [inline content/articles/creating-a-hand-gesture-volume-controller-using-python-and-pycharm/index.md:159]:
```suggestion
- In the code above, we use the first for loop to interact with each hand in the results. We use the second for loop to get the `id` (id number) and `lm` (landmark information) for each hand landmark. The landmark information will give us the `x` and `y` coordinates. The id number is the number assigned to the various hand points.
```

File: content/articles/creating-a-hand-gesture-volume-controller-using-python-and-pycharm/index.md [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+ 
+### Step 3: Accessing the speaker using pycaw
+```python
+devices = AudioUtilities.GetSpeakers()
+interface = devices.Activate(IAudioEndpointVolume._iid_, CLSCTX_ALL, None)
+volume = cast(interface, POINTER(IAudioEndpointVolume))
+```
+These are the initializations we need for `pycaw` to run smoothly. The developer provides this library together with the initializations. We are not going to change anything. You can find the documentation [here](https://github.com/AndreMiras/pycaw).
+
+### Step4: Finding the volume range between the minimum and maximum volume
```
Comment (Bayler) [inline content/articles/creating-a-hand-gesture-volume-controller-using-python-and-pycharm/index.md:124]:
```suggestion
### Step 4: Finding the volume range between the minimum and maximum volume
```

File: content/articles/creating-a-hand-gesture-volume-controller-using-python-and-pycharm/index.md [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+
+ Lets now detect, initialize and configure the hands in the video input.
+
+### Step 2: Detecting, initializing, and configuring the hands
+```python
+mpHands = mp.solutions.hands
+hands = mpHands.Hands()
+mpDraw = mp.solutions.drawing_utils
+```
+In the above code, we are calling on the `mediapipe` hand module to detect the hands from the video input we got from our primary camera. `MpHands.Hands()` then completes the initialization and configuration of the detected hands. We finally draw the *connections* and *landmarks* on the detected hand using `mp.solutions.drawing_utils`.
```
Comment (Bayler) [inline content/articles/creating-a-hand-gesture-volume-controller-using-python-and-pycharm/index.md:114]: Please replace, 'In the above code', with 'In the code above'. It sounds more grammatically correct. Go through your entire tutorial and edit accordingly.

File: content/articles/creating-a-hand-gesture-volume-controller-using-python-and-pycharm/index.md [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+import numpy as np
+```
+In the above code segment, we are importing each library we installed in our project.
+
+```python
+cap = cv2.VideoCapture(0) 
+```
+We then get the video input from our computer's primary camera. If you are using any other camera, replace the number `0` with that of the camera you are using.
+
+ Lets now detect, initialize and configure the hands in the video input.
```
Comment (Bayler) [inline content/articles/creating-a-hand-gesture-volume-controller-using-python-and-pycharm/index.md:106]:
```suggestion
Let's now detect, initialize and configure the hands in the video input.
```

**Comment by leeseshawn** [comment]
@Bayler 
 Done working on the changes.

**Comment by ahmadmardeni1** [comment]
Good work team. @Bayler @leeseshawn 

Welcome to Section EngEd and congrats on your first article. I really recommend you take the tech writing course provided by Google: https://developers.google.com/tech-writing

**Pull request merged**